crates/turbopack-ecmascript/src/analyzer/mod.rs

```diff
@@ -1007,58 +1214,375 @@ impl JsValue {
             }
         }
     }
+}
 
+// Unknown management
+impl JsValue {
+    /// Convert the value into unknown with a specific reason.
     pub fn make_unknown(&mut self, reason: &'static str) {
         *self = JsValue::Unknown(Some(Arc::new(take(self))), reason);
     }
 
+    /// Convert the value into unknown with a specific reason, but don't retain
+    /// the original value.
     pub fn make_unknown_without_content(&mut self, reason: &'static str) {
         *self = JsValue::Unknown(None, reason);
     }
 
-    pub fn has_placeholder(&self) -> bool {
-        match self {
-            // These are leafs and not placeholders
-            JsValue::Constant(_)
-            | JsValue::Url(_)
-            | JsValue::WellKnownObject(_)
-            | JsValue::WellKnownFunction(_)
-            | JsValue::Unknown(_, _)
-            | JsValue::Function(..) => false,
-
-            // These must be optimized reduced if they don't contain placeholders
-            // So when we see them, they contain placeholders
-            JsValue::Call(..) | JsValue::MemberCall(..) | JsValue::Member(..) => true,
+    /// Make all nested operations unknown when the value is an operation.
+    pub fn make_nested_operations_unknown(&mut self) -> bool {
+        fn inner(this: &mut JsValue) -> bool {
+            if matches!(this.meta_type(), JsValueMetaKind::Operation) {
+                this.make_unknown("nested operation");
+                true
+            } else {
+                this.for_each_children_mut(&mut inner)
+            }
+        }
+        if matches!(self.meta_type(), JsValueMetaKind::Operation) {
+            self.for_each_children_mut(&mut inner)
+        } else {
+            false
+        }
+    }
+}
 
-            // These are nested structures, where we look into children
-            // to see placeholders
-            JsValue::Array(..)
-            | JsValue::Object(..)
-            | JsValue::Alternatives(..)
-            | JsValue::Concat(..)
-            | JsValue::Add(..) => {
+// Placeholder management
+impl JsValue {
+    /// Returns true if the value contains or is a placeholder.
+    pub fn has_placeholder(&self) -> bool {
+        match self.meta_type() {
+            JsValueMetaKind::Placeholder => true,
+            JsValueMetaKind::Leaf => false,
+            JsValueMetaKind::Nested | JsValueMetaKind::Operation => {
                 let mut result = false;
                 self.for_each_children(&mut |child| {
```

Comment:
Unrelated to this PR, but this should be short-circuitable, using an `any` instead of a `for_each`.